### ISSUE TITLE
Caching the orbits of massive bodies with updateable massless bodies

### ISSUE THREAD

**Issue by expenses** [Canleskis/particular#11]
Hi! I just discovered your crate while looking for a n-body solver and it looks fantastic so far!

I'm interested in write a space sim game that uses orbital mechanics similar to Kerbal Space Program (which just has 2-body physics) or Children of a Dead Earth (which has n-body). One thing for performance that I'd like to see if the ability to cache the orbits of all the bodies with mass, e.g. planets, moons, comets etc for a certain amount of time ahead of time, and then using those for faster calculation of the orbits of effectively massless bodies like a spaceship.

My understanding is that this isn't possible in the crate so far. I'm totally happy to have a go at implementing something like this, but do you have any insights into this idea? It is worth it in a performance/storage trade-off?

**Comment by Canleskis** [comment]
Hey, thanks for the kind words!

This sounds like an interesting problem. Stripping down the idea, what we would actually need is the ability to add a massless body and get its acceleration without recomputing the acceleration of all the bodies.

The easiest way is probably to use two `ComputeMethod`s, one for the initial calculations ahead of time, and another one for the faster calculation of the massless bodies. The first one could use any of the built-in ones, but the second one does not exist yet.

The second `ComputeMethod` would need to only compute the acceleration for the massless bodies using the massive bodies. There are multiple ways you could achieve this, but the most efficient is probably storing the massive bodies in an instance of the struct to be used in `compute()`.

```rust
pub struct MasslessAccelerations {
    massive: Vec<(Vec2, f32)>,
}

impl ComputeMethod<Vec2, f32> for MasslessAccelerations {
    fn compute(&mut self, particles: &[(Vec2, f32)]) -> Vec<Vec2> {
        particles
            .iter()
            .map(|&(position1, _)| {
                self.massive
                    .iter()
                    .fold(Vec2::ZERO, |acceleration, &(position2, mass2)| {
                        let dir = position2 - position1;
                        let mag_2 = dir.length_squared();

                        let grav_acc = if mag_2 != 0.0 {
                            dir * mass2 / (mag_2 * mag_2.sqrt())
                        } else {
                            dir
                        };

                        acceleration + grav_acc
                    })
            })
            .collect()
    }
}
```

This is an example in 2D (Vec2), but you could probably reuse how built-in `ComputeMethod`s are made generic to make this more general.
Using it would simply require you to use the `accelerations` or `map_accelerations` method on an iterator of your massless bodies.
If you plan on using BarnesHut this is however not yet possible, but it is just a case of making the tree and its traits public to be able to be used in user-defined `ComputeMethod`s, which will be done once I complete their documentation.

Hope this helps and that I understood your problem correctly, and if you have more questions I'd be happy to answer them!

_Edit_: To sum up, because I realised it might not be completely clear how this solves the problem:

Using one of the built-in `ComputeMethod`, you will have to compute the acceleration of your massive bodies for the given amount of time you want to cache the data for and use that acceleration to integrate the position of the bodies. 

You'll cache the state of these bodies for each step. The data structure for this would probably need to be a Vec<Vec<_>>, the first vector's entries corresponding to each step, and the nested vector holding the data for the bodies.

Then, for a given step, you can use the corresponding entry in the vector to build a `MasslessAccelerations` instance and compute the acceleration of the massless bodies.

**Comment by Canleskis** [comment]
I wasn't fully satisfied of the answer I gave you because I felt that Particular wasn't flexible enough for a case like yours, where you need more control over what is sent over in a given `ComputeMethod` to be able to reuse existing ones. So I spent some time thinking of a solution to improve such cases and I came up with a solution that I think facilitates use case like yours.

I also added a `sequential::BruteForceAlt` `ComputeMethod` that is more easily reused than `sequential::BruteForce` because of how it works internally, but is slower with massive particles.

In short, `ComputeMethod`s are now generic over a type bound to `ParticleStorage`. This allows for more flexibility and prevents some possible misuses. Now, for your specific use case, you can simply have:
```rust
pub struct MasslessAccelerations {
    massive: Vec<(Vec2, f32)>,
}

impl super::ComputeMethod<Vec<(Vec2, f32)>> for MasslessAccelerations {
    fn compute(&mut self, storage: Vec<(Vec2, f32)>) -> Vec<Vec2> {
        BruteForceAlt.compute(WithMassive {
            particles: storage,
            massive: std::mem::take(&mut self.massive),
        })
    }
}
```
This would be used by iterating over the massless bodies and calling one of Particular's iterator method.

Another solution would be this:
```rust
pub struct MasslessAccelerations2;

impl super::ComputeMethod<WithMassive<glam::Vec2, f32>> for MasslessAccelerations2 {
    fn compute(&mut self, mut storage: WithMassive<glam::Vec2, f32>) -> Vec<glam::Vec2> {
        storage.particles.retain(|(_, mu)| *mu == 0.0);
        BruteForceAlt.compute(storage)
    }
}
```
Whilst this looks simpler, you have to be careful when using it. This requires an iterator with the massless bodies first, then the massive bodies (using f.e. chain).

I'd recommend using the first solution as it is clearer what it is doing.
There are probably other solutions since this is quite flexible, but I didn't want to delve into it too much.

I'll try to have this update up on crates.io next week after more thorough testing.

**Comment by Canleskis** [comment]
I can close this now that 13c4ead is released as part of [0.6.0](https://github.com/Canleskis/particular/compare/v0.5.2...v0.6.0). 
The release took longer than expected because this was part of a large overhaul of the backend API to try to make it more robust and flexible. This allowed for the addition of `BruteForceSIMD` compute methods notably.

Here is a generic solution using Particular's provided vector abstractions:

```rust
use particular::{
    algorithms::{internal, MassiveAffected, ParticleSet},
    prelude::*,
};

pub struct MasslessAccelerations<'a, P> {
    massive: &'a [P],
}

impl<T, S, P> ComputeMethod<ParticleSet<T, S>, P::Vector> for MasslessAccelerations<'_, P>
where
    S: internal::Scalar,
    T: internal::Vector<Scalar = S>,
    P: Particle<Scalar = S>,
    P::Vector: internal::IntoVectorArray<T::Array, Vector = T>,
{
    type Output = Vec<P::Vector>;

    fn compute(self, storage: ParticleSet<T, S>) -> Self::Output {
        sequential::BruteForce.compute(MassiveAffected {
            affected: storage.0,
            massive: self
                .massive
                .iter()
                .map(|p| p.point_mass().into_internal())
                .collect(),
        })
    }
}
```
Using it simply requires calling `accelerations` on an iterator of particles you want to compute the acceleration for, passing in an instance of `MasslessAccelerations` with a slice of the particles you want to be considered massive.

Note that there are other ways this could be achieved, such as using a chained iterator of massive and affected particles, and have the compute method only computing and returning the acceleration for the affected particles. 
You could also create your own storage that's a wrapper around `MassiveAffected` which populates the `affected` vector with massless particles only.

For this specific problem, you should be able to reuse any of the compute method that are implemented using `MassiveAffected` for their storage.
